doc/dox/programs/pw-loopback.1.md
\page page_man_pw-loopback_1 pw-loopback

PipeWire loopback client

# SYNOPSIS

**pw-loopback** \[*options*\]

# DESCRIPTION

The *pw-loopback* program is a PipeWire client that uses the PipeWire
loopback module to create loopback nodes, with configuration given via
the command-line options.

# OPTIONS

\par -h | \--help
Show help.

\par -r | \--remote=NAME
The name of the *remote* instance to connect to. If left unspecified, a
connection is made to the default PipeWire instance.

\par -n | \--name=NAME
Name of the loopback node

\par -g | \--group=NAME
Name of the loopback node group

\par -c | \--channels=NUMBER
Number of channels to provide

\par -m | \--channel-map=MAP
Channel map (default `[ FL, FR ]`)

\par -l | \--latency=LATENCY
Desired latency in ms

\par -d | \--delay=DELAY
Added delay in seconds (floating point allowed)

\par -C | \--capture=TARGET
Target device to capture from

\par -P | \--playback=TARGET
Target device to play to

\par -i | \--capture-props=PROPS
Wanted properties of capture node (in JSON)

\par -o | \--playback-props=PROPS
Wanted properties of playback node (in JSON)

# AUTHORS

The PipeWire Developers <$(PACKAGE_BUGREPORT)>;
PipeWire is available from <$(PACKAGE_URL)>

# SEE ALSO

\ref page_man_pipewire_1 "pipewire(1)",
\ref page_man_pw-cat_1 "pw-cat(1)",
**pactl(1)**

Other ways to create loopback nodes are adding the loopback module in
the configuration of a PipeWire daemon, or loading the loopback module
using Pulseaudio commands (`pactl load-module module-loopback ...`).


NEWS
# PipeWire 1.5.85 (2026-01-19)

This is the fifth and hopefully last 1.6 release candidate that
is API and ABI compatible with previous 1.4.x, 1.2.x and 1.0.x
releases.

## Highlights
  - Fix errors in older clients that blindly mmap buffer data.
  - ALSA channel mapping should work in more cases.
  - Fix an overflow in the resampler with some sample rates.
  - Device ID negotiation for DMABUF was added.
  - The channelmixer can now also downmix most REAR and TOP
    channel positions.
  - Various small fixes and improvements.


## PipeWire
  - Avoid doing flushing invoke in pw-stream. (#5010)
  - Fix stream time delay calculations in some cases. (#4957)
  - Avoid suspending a node when the ports are busy.

## Modules
  - Mode AVB module updates.
  - Remove MAPPABLE flag from memptr data in client-node. This avoids
    errors in older clients that blindly map buffers. (#5003)
  - Don't leak the filter-chain module in module-parametric-equalizer.
    (#5045)
  - make it possible to run pulse tunnel sink and source with the same
    name. (#5079)

## SPA
  - The delay filter in filter-graph now has feedback and feedforward
    options to make it possible to implement comb and allpass filters.
  - The resampler can be compiled with a custom default quality.
  - Firewire latency is scaled correctly now. (#4977)
  - The mixer will only passthrough dynamic data.
  - Fix filter-graph state in audioconvert after flush.
  - Set the channel map in ALSA nodes to tell ALSA our channel layout
    and avoid wrong channels.
  - Fix AVX2 function, make sure they don't run on AVX-only CPUs.
    (#5072)
  - Fix an overflow and crash when too many phases are used in the
    resampler. (#5073)
  - Add some more channel downmix positions, mostly REAR and TOP
    channels.

## pulse-server
  - Add a message to enable mono mixdown.

## Tools
  - Support filenames in raw mode in pw-cat. (#5012)
  - Enable clipping in spa-resample.

## examples
  - Update device ID negotiation examples.

## Docs
  - Document the resampler properties better.


Older versions:

# PipeWire 1.5.84 (2025-11-27)

This is the fourth 1.6 release candidate that is API and ABI
compatible with previous 1.4.x, 1.2.x and 1.0.x releases.

Changes since the last pre-release:

## Highlights
  - Capabilities wer added to improve negotiation over links.
  - The audio resampler now has a configurable window function to better
    tune the resampler quality. A kaiser and blackman window was added
    and the default parameters were tuned.
  - Various small fixes and improvements.

## PipeWire
  - Capabilities and PeerCapabilities were added to exchange key/value
    pairs between consumer and producer right after a link is made. This
    can be used to detect how the negotiation of formats and buffers
    should be done.

## Modules
  - Avoid segfaults in RTP source. (#4970) 
  - The AVB module has seen some improvements.

## Pulse-server
  - @NONE@ can now be used to clear the default sink/source.

## SPA
  - Support longer convolver filenames and also support inline
    IRs.
  - The audio resampler window function is now selectable and
    configurable. A kaiser window and blackman window was added
    and the default qualities were tweaked to improve quality.
  - The filter-graph convolver latency is now set by default to
    something more sensible. (0 by default and N/2 for hilbert).
    (#4980)

## Bluetooth
  - Better xrun and error handling for iso streams.
  - The +CNUM reply was fixed.
  - The CIEC call status was fixed. (#1744)
  - Add BAP context metadata to improve compatibility.
  - Improve compatiblity with Creative Zen Hybrid Pro by releasing
    transports simultaneously.

# PipeWire 1.5.83 (2025-11-06)

This is the third 1.6 release candidate that is API and ABI
compatible with previous 1.4.x, 1.2.x and 1.0.x releases.

Changes since the last pre-release:

## Highlights
  - Include the NEWS and updated version number.

# PipeWire 1.5.82 (2025-11-06)

This is the second 1.6 release candidate that is API and ABI
compatible with previous 1.4.x, 1.2.x and 1.0.x releases.

Changes since the last pre-release:

## Highlights
  - The max channel limit is now a compile time option.
  - The SAP and RTP module have seen some robustness improvements.
  - Add audio.layout propperty.
  - Cleanups to the code here and there.

## PipeWire
  - Handle Tags more like Latency with a NULL param when no ports are linked
    and some sort of (empty) Tag when the ports are linked.

## Modules
  - Improve the echo-cancel module to keep the streams more aligned
    and cause less latency.
  - Improve format parsing errors in most modules.
  - The RTP module now has extra code for better network robustness, including
    cases when network interfaces are not yet up and running, and multicast
    sockets are silently kicked out of IGMP groups.
  - The direct timestamp mode in the RTP module was effectively broken and is
    now fixed.
  - Add support for audio.layout.
  - Add multichannel support to ROC.

## SPA
  - Rework the maximum number of channel handling. Because this is a
    potential ABI break, it is now a compile time option with new
    functions to handle more than the previous 64 channels.
  - The 64 channel limit was removed from the noise shaper.
  - spa_strbuf is used in more places instead of custom snprintf code.
  - The volume ramp code was simplified.
  - The driver node now has properties to configure the clock.
  - The adapter will try to renegotiate when the driver changes.
  - Fix relaxed array parsing with od number of elements. (#4944)
  - audio.layout was added to set the channel positions to some
    predefined layouts.
  - Added more POD choice checks to ensure the right amount of values
    are present in the choice.
  - Fix __has_attribute usage. (#4962)
  - Thread RESET_ON_FORK is now disabled for JACK application so that
    forking will preserve any real-time thread priorities, like JACK.
    (#4966)

src/tools/pw-loopback.c
/* PipeWire */
/* SPDX-FileCopyrightText: Copyright © 2021 Wim Taymans <wim.taymans@gmail.com> */
/* SPDX-License-Identifier: MIT */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <getopt.h>
#include <limits.h>
#include <math.h>
#include <locale.h>

#include <spa/utils/result.h>
#include <spa/pod/builder.h>
#include <spa/param/audio/format-utils.h>
#include <spa/param/audio/raw.h>
#include <spa/utils/json.h>
#include <spa/debug/file.h>

#include <pipewire/pipewire.h>
#include <pipewire/impl.h>

#define DEFAULT_RATE		48000
#define DEFAULT_CHANNELS	2
#define DEFAULT_CHANNEL_MAP	"[ FL, FR ]"

struct data {
	struct pw_main_loop *loop;
	struct pw_context *context;

	struct pw_impl_module *module;
	struct spa_hook module_listener;

	const char *opt_node_name;
	const char *opt_group_name;
	const char *opt_channel_map;

	uint32_t channels;
	uint32_t latency;
	float delay;

	struct pw_properties *capture_props;
	struct pw_properties *playback_props;
};

static void do_quit(void *data, int signal_number)
{
	struct data *d = data;
	pw_main_loop_quit(d->loop);
}

static void module_destroy(void *data)
{
	struct data *d = data;
	spa_hook_remove(&d->module_listener);
	d->module = NULL;
	pw_main_loop_quit(d->loop);
}

static const struct pw_impl_module_events module_events = {
	PW_VERSION_IMPL_MODULE_EVENTS,
	.destroy = module_destroy
};


static void show_help(struct data *data, const char *name, bool error)
{
	fprintf(error ? stderr : stdout, "%s [options]\n"
		"  -h, --help                            Show this help\n"
		"      --version                         Show version\n"
		"  -r, --remote                          Remote daemon name\n"
		"  -n, --name                            Node name (default '%s')\n"
		"  -g, --group                           Node group (default '%s')\n"
		"  -c, --channels                        Number of channels (default %d)\n"
		"  -m, --channel-map                     Channel map (default '%s')\n"
		"  -l, --latency                         Desired latency in ms\n"
		"  -d, --delay                           Desired delay in float s\n"
		"  -C  --capture                         Capture source to connect to (name or serial)\n"
		"  -i  --capture-props                   Capture stream properties\n"
		"  -P  --playback                        Playback sink to connect to (name or serial)\n"
		"  -o  --playback-props                  Playback stream properties\n",
		name,
		data->opt_node_name,
		data->opt_group_name,
		data->channels,
		data->opt_channel_map);
}

int main(int argc, char *argv[])
{
	struct data data = { 0 };
	struct pw_loop *l;
	const char *opt_remote = NULL;
	char cname[256], value[256];
	char *args;
	size_t size;
	FILE *f;
	static const struct option long_options[] = {
		{ "help",		no_argument,		NULL, 'h' },
		{ "version",		no_argument,		NULL, 'V' },
		{ "remote",		required_argument,	NULL, 'r' },
		{ "group",		required_argument,	NULL, 'g' },
		{ "name",		required_argument,	NULL, 'n' },
		{ "channels",		required_argument,	NULL, 'c' },
		{ "channel-map",	required_argument,	NULL, 'm' },
		{ "latency",		required_argument,	NULL, 'l' },
		{ "delay",		required_argument,	NULL, 'd' },
		{ "capture",		required_argument,	NULL, 'C' },
		{ "playback",		required_argument,	NULL, 'P' },
		{ "capture-props",	required_argument,	NULL, 'i' },
		{ "playback-props",	required_argument,	NULL, 'o' },
		{ NULL, 0, NULL, 0}
	};
	int c, res = -1;
	struct spa_error_location loc;

	setlocale(LC_ALL, "");
	pw_init(&argc, &argv);

	data.channels = DEFAULT_CHANNELS;
	data.opt_channel_map = DEFAULT_CHANNEL_MAP;
	data.opt_group_name = pw_get_client_name();
	if (snprintf(cname, sizeof(cname), "%s-%zd", argv[0], (size_t) getpid()) > 0)
		data.opt_group_name = cname;
	data.opt_node_name = data.opt_group_name;

	data.capture_props = pw_properties_new(NULL, NULL);
	data.playback_props = pw_properties_new(NULL, NULL);
	if (data.capture_props == NULL || data.playback_props == NULL) {
		fprintf(stderr, "can't create properties: %m\n");
		goto exit;
	}

	while ((c = getopt_long(argc, argv, "hVr:n:g:c:m:l:d:C:P:i:o:", long_options, NULL)) != -1) {
		switch (c) {
		case 'h':
			show_help(&data, argv[0], false);
			return 0;
		case 'V':
			printf("%s\n"
				"Compiled with libpipewire %s\n"
				"Linked with libpipewire %s\n",
				argv[0],
				pw_get_headers_version(),
				pw_get_library_version());
			return 0;
		case 'r':
			opt_remote = optarg;
			break;
		case 'n':
			data.opt_node_name = optarg;
			break;
		case 'g':
			data.opt_group_name = optarg;
			break;
		case 'c':
			data.channels = atoi(optarg);
			break;
		case 'm':
			data.opt_channel_map = optarg;
			break;
		case 'l':
			data.latency = atoi(optarg) * DEFAULT_RATE / SPA_MSEC_PER_SEC;
			break;
		case 'd':
			if (!spa_atof(optarg, &data.delay))
				data.delay = (float)atof(optarg);
			break;
		case 'C':
			pw_properties_set(data.capture_props, PW_KEY_TARGET_OBJECT, optarg);
			break;
		case 'P':
			pw_properties_set(data.playback_props, PW_KEY_TARGET_OBJECT, optarg);
			break;
		case 'i':
			if (pw_properties_update_string_checked(data.capture_props,
						optarg, strlen(optarg), &loc) < 0) {
				spa_debug_file_error_location(stderr, &loc,
						"error: syntax error in --capture-props: %s",
						loc.reason);
				return -1;
			}
			break;
		case 'o':
			if (pw_properties_update_string_checked(data.playback_props,
						optarg, strlen(optarg), &loc) < 0) {
				spa_debug_file_error_location(stderr, &loc,
						"error: syntax error in --playback-props: %s",
						loc.reason);
				return -1;
			}
			break;
		default:
			show_help(&data, argv[0], true);
			return -1;
		}
	}

	data.loop = pw_main_loop_new(NULL);
	if (data.loop == NULL) {
		fprintf(stderr, "can't create main loop: %m\n");
		goto exit;
	}

	l = pw_main_loop_get_loop(data.loop);
	pw_loop_add_signal(l, SIGINT, do_quit, &data);
	pw_loop_add_signal(l, SIGTERM, do_quit, &data);

	data.context = pw_context_new(l, NULL, 0);
	if (data.context == NULL) {
		fprintf(stderr, "can't create context: %m\n");
		goto exit;
	}


	if ((f = open_memstream(&args, &size)) == NULL) {
		fprintf(stderr, "can't open memstream: %m\n");
		goto exit;
	}

	fprintf(f, "{");

	if (opt_remote != NULL)
		fprintf(f, " remote.name = \"%s\"", opt_remote);
	if (data.latency != 0)
		fprintf(f, " node.latency = %u/%u", data.latency, DEFAULT_RATE);
	if (data.delay != 0.0f)
		fprintf(f, " target.delay.sec = %s",
				spa_json_format_float(value, sizeof(value), data.delay));
	if (data.channels != 0)
		fprintf(f, " audio.channels = %u", data.channels);
	if (data.opt_channel_map != NULL)
		fprintf(f, " audio.position = %s", data.opt_channel_map);
	if (data.opt_node_name != NULL)
		fprintf(f, " node.name = %s", data.opt_node_name);

	if (data.opt_group_name != NULL) {
		pw_properties_set(data.capture_props, PW_KEY_NODE_GROUP, data.opt_group_name);
		pw_properties_set(data.playback_props, PW_KEY_NODE_GROUP, data.opt_group_name);
	}

	fprintf(f, " capture.props = {");
	pw_properties_serialize_dict(f, &data.capture_props->dict, 0);
	fprintf(f, " } playback.props = {");
	pw_properties_serialize_dict(f, &data.playback_props->dict, 0);
	fprintf(f, " } }");
	fclose(f);

	pw_log_info("loading module with %s", args);

	data.module = pw_context_load_module(data.context,
			"libpipewire-module-loopback", args,
			NULL);
	free(args);

	if (data.module == NULL) {
		fprintf(stderr, "can't load module: %m\n");
		goto exit;
	}

	pw_impl_module_add_listener(data.module,
			&data.module_listener, &module_events, &data);

	pw_main_loop_run(data.loop);

	res = 0;
exit:
	if (data.module)
		pw_impl_module_destroy(data.module);
	if (data.context)
		pw_context_destroy(data.context);
	if (data.loop)
		pw_main_loop_destroy(data.loop);
	pw_properties_free(data.capture_props);
	pw_properties_free(data.playback_props);
	pw_deinit();

	return res;
}
